src/Samples/Sarif.WorkItems.Sample/Program.cs

```diff
@@ -5,112 +5,83 @@
         {
-            NonFunctioningEndToEndApiUsageExample();
-            return 0;
-        }
-
-        private static void NonFunctioningEndToEndApiUsageExample()
-        {
+            // Mock up of end-to-end API use.
+            
             // 1. Retrieve the current SARIF and its baseline equivalent
-            SarifLog baselineSarif = RetrieveBaselineSarif();
-            SarifLog currentSarif = RetrieveCurrentSarif();
+            SarifLog baselineSarif = RetrieveBaselineSarif(args[0]);
+            SarifLog currentSarif = RetrieveCurrentSarif(args[1]);
+            string outputFilePath = args[2];
 
-            // 2. Match the SARIF files
+            // 2. Given a baseline and current sarif, produce a 'matched' log
             ISarifLogBaseliner resultMatcher = SarifLogBaselinerFactory.CreateSarifLogBaseliner(SarifBaselineType.Standard);
 
-            SarifLog matchedLog = new SarifLog
-            {
-                Runs = new[]
-                {
-                    resultMatcher.CreateBaselinedRun(baselineSarif.Runs[0], currentSarif.Runs[0])
-                }
-            };
+            SarifLog matchedSarif = resultMatcher.CreateBaselinedSarifLog(
+                baselineSarif, 
+                currentSarif);
 
-            // 3. Make sure everything has a GUID and that we have a map back to each result from it
-            var guidToMatchedResultMap = new Dictionary<string, Result>();
-            foreach (Result result in matchedLog.Runs[0].Results)
-            {
-                if (result.Guid == null)
-                {
-                    result.Guid = Guid.NewGuid().ToString();
-                }
-                guidToMatchedResultMap[result.Guid] = result;
-            }
+            // 3. We need a map from each result guid to the matched SARIF result.
+            //    As bugs are filed against refactored files, we use the guid to 
+            //    link to the original file in order to populate the WIT URI.
 
-            var adoClient = new AzureDevOpsClient();
-            var filingTarget = new AzureDevOpsFilingTarget(adoClient);
+            var guidMappingVisitor = new GuidMappingVisitor();
+            guidMappingVisitor.Visit(matchedSarif);
 
-            var workItemFiler = new WorkItemFiler(
-                filingTarget,
-                filteringStrategy: new NewResultsFilteringStrategy(),
-                groupingStrategy: new OneResultPerWorkItemGroupingStrategy());
+            // 4. Now we need to refactor the log file into one or more log files, 
+            // each of which will drive the creation of a single bug.
 
-            string logFilePath = Path.GetTempFileName();
-            File.WriteAllText(logFilePath, JsonConvert.SerializeObject(matchedLog));
+            IEnumerable<SarifLog> sarifLogsToFile = SarifSplitter.Split(
+                sarifLog: matchedSarif,
+                filterPredicate: (result) => { return result.BaselineState == BaselineState.New; },
```

Comment:
>filterPredicate [](start = 16, length = 15)

We don't want users to have to write their own predicates (I mean they can, but they shouldn't have to). So we should supply a set of predicates (or strategies, see next comment) so they can write either
```
filteringPredicate = FilteringPredicates.NewResults    // a static property of delegate type
```
or
```
filteringStrategy = new NewResultsFilteringStrategy()
```
or
```
filterStrategy = FilteringStrategies.NewResults // a pre-built NewResultsFilteringStrategy object
```

Reply:
Yes.

---
In reply to: [307084390](https://github.com/microsoft/sarif-sdk/pull/1589#discussion_r307084390) [](ancestors = 307084390)